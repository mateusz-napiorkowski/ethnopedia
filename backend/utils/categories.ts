import CollectionCollection from "../models/collection";
import { collectionCategory, artworkCategory } from "./interfaces"

export const hasValidCategoryFormat = (categories: Array<collectionCategory>, isRootArray = true): boolean => {
    if(isRootArray && !categories.length)
        return false
    return categories.every(category => 
        category?.name && 
        Array.isArray(category?.subcategories) && 
        hasValidCategoryFormat(category.subcategories, false)
    );
};

export const artworkCategoriesHaveValidFormat = (artworkCategories: Array<artworkCategory>, collectionCategories: Array<collectionCategory>, isRootArray = true): boolean => {
    if (isRootArray && (artworkCategories.length == 0 || collectionCategories.length == 0)) return false;
    if (artworkCategories.length !== collectionCategories.length) return false;

    return collectionCategories.every((category, index) => {
        const artworkCategory = artworkCategories[index];

        return (
            category.name === artworkCategory.name &&
            !!artworkCategory.value &&
            artworkCategoriesHaveValidFormat(artworkCategory.subcategories, category.subcategories, false)
        );
    });
};

export const isValidCollectionCategoryStructureForCollectionUpdate = (artworkSubcategories: Array<artworkCategory>, collectionSubcategories: Array<collectionCategory>): boolean => {
    if(artworkSubcategories.length > collectionSubcategories.length) return false
    return artworkSubcategories.every((artworkCategory, index) =>
        isValidCollectionCategoryStructureForCollectionUpdate(
          artworkCategory.subcategories,
          collectionSubcategories[index].subcategories
        )
      );
}

const getNestedCategories = ((prefix: string, subcategories: Array<artworkCategory>) => {
    const nestedCategories: Array<string> = []
    for(const subcategory of subcategories) {
        nestedCategories.push(`${prefix}.${subcategory.name}`)
        nestedCategories.push(...getNestedCategories(`${prefix}.${subcategory.name}`, subcategory.subcategories))
    }
    return nestedCategories
})

export const getAllCategories = async (collectionName: string) => {
    try {
        const collections = await CollectionCollection.find({name: collectionName}).exec()
        if(collections.length !== 1)
            throw new Error("Collection not found")
        const allCategories: Array<string> = []
        for(const category of collections[0].categories) {
            allCategories.push(category.name)
            allCategories.push(...getNestedCategories(`${category.name}`, category.subcategories))
        }           
        return allCategories
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Collection not found`)
            throw new Error("Collection not found")
        else
            throw new Error("Database unavailable")
    }
}

export const transformCategoriesArrayToCategoriesObject = (categoriesArray: Array<string>, depth = 1) => {
    const categories: Array<any> = []
    categoriesArray.forEach(field => {
        const allSubcategoriesNames = categoriesArray
            .filter(columnName => (columnName.startsWith(`${field}.`)))
            .map(item => item.split('.').slice(1).join('.'));
        if(!field.includes("."))
            categories.push({
                name: field,
                subcategories: transformCategoriesArrayToCategoriesObject(allSubcategoriesNames, depth + 1)
            })        
    })
    return categories
}

export const findMissingParentCategories = ((categoriesArray: Array<string>) => {
    const missingCategories: Array<string> = []
    categoriesArray.forEach((category: string) => {
        const parentCategory = category.split(".").slice(0, -1).join(".")
        if(category.includes(".") && !categoriesArray.includes(parentCategory))
            missingCategories.push(parentCategory)
    });
    return missingCategories
})