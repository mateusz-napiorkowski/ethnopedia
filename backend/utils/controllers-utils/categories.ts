export const getNestedCategories = ((prefix: string, subcategories: any) => {
    const nestedCategories: Array<string> = []
    if(subcategories !== undefined) {
        for(const subcategory of subcategories){
            nestedCategories.push(`${prefix}${subcategory.name}`)
            nestedCategories.push(...getNestedCategories(`${prefix}${subcategory.name}.`, subcategory.subcategories))
        }
    }
    return nestedCategories
})