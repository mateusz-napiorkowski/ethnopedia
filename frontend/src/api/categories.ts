import axios from "axios"
import { API_URL } from "../config"

export const getCategories = async (id: string) => {
    const response = await axios.get(`${API_URL}v1/categories/${id}`)
    return response.data
}
