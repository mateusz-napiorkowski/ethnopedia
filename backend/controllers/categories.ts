import { Request, Response } from "express"
import { getAllCategories } from "../utils/categories"

export const getCollectionCategories = async (req: Request, res: Response) => {
    try {
        const collectionName = Buffer.from(req.params.collectionName, 'latin1').toString("utf-8")
        const categories = await getAllCategories(collectionName)
        res.status(200).json({categories: categories})
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Collection not found`)
            res.status(404).json({ error: err.message })
        else
            res.status(503).json({ error: `Database unavailable` })
    }
    
}