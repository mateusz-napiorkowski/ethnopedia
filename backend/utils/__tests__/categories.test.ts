import {describe, expect, test, jest, beforeEach} from "@jest/globals"
import { getAllCategories } from "../categories"

const mockArtworkFind = jest.fn()
jest.mock('../../models/artwork', () => ({
    find: () => mockArtworkFind()
}))

const mockCollectionFind = jest.fn()
jest.mock('../../models/collection', () => ({
    find: () => mockCollectionFind()
}))

describe('categories util functions tests', () => {
    beforeEach(() => {
        jest.resetAllMocks()
    })

    test.each([
        {
            testName: 'getAllColletions test - artworks present in collection',
            artworkFind: () => {return {exec: () => Promise.resolve([
                {
                  _id: "6717d4c0666e8575d873ee69",
                  createdAt: '2024-10-22T20:12:12.209Z',
                  updatedAt: '2024-10-22T20:12:12.209Z',
                  __v: 0,
                  categories: [
                    {
                      name: 'Tytuł',
                      values: [ 'testowy' ],
                      subcategories: [
                        {
                          name: 'Podtytuł',
                          values: [ 'podtytuł testowy' ],
                          subcategories: [
                            {
                              name: 'Podpodtytuł',
                              values: [ 'podpodtytuł testowy' ]
                            }
                          ]
                        }
                      ]
                    },
                    { name: 'Artyści', values: [ 'testowi' ], subcategories: [] },
                    {
                      name: 'Rok',
                      values: [ '966' ],
                      subcategories: [
                        {
                          name: 'Miesiąc',
                          values: [ '12' ],
                          subcategories: [ { name: 'Dzień', values: [ '13' ] } ]
                        }
                      ]
                    }
                  ],
                  collectionName: 'collection'
                },
                {
                  _id: "6718078ad4821e244dd54b84",
                  createdAt: '2024-10-22T20:14:13.773Z',
                  updatedAt: '2024-10-22T20:14:13.773Z',
                  __v: 0,
                  categories: [
                    {
                      name: 'Tytuł',
                      values: [ 'testowy 2' ],
                      subcategories: [ { name: 'Znaczenie tytułu', values: [ 'testowe' ] } ]
                    },
                    {
                      name: 'Artyści',
                      values: [ 'artysta testowy' ],
                      subcategories: []
                    },
                    {
                      name: 'Rok',
                      values: [ '955' ],
                      subcategories: [
                        {
                          name: 'Kwartał',
                          values: [ 'III' ],
                          subcategories: [
                            {
                              name: 'Miesiąc',
                              values: [ 'Wrzesień' ],
                              subcategories: [ { name: 'Dzień', values: [ '1' ] } ]
                            }
                          ]
                        }
                      ]
                    }
                  ],
                  collectionName: 'collection'
                }
              ])}},
        },
        {
            testName: 'getAllColletions test - no artworks in collection',
            artworkFind: () => {return {exec: () => Promise.resolve([])}},
        },
    ])(`$testName`,
        async ({artworkFind}) => {
            mockCollectionFind.mockImplementation(() => {return {exec: () => Promise.resolve([
                {
                    _id: "6717d46c666e8575d873ee57",
                    name: 'collection',
                    description: 'collection description',
                    __v: 0
                }
            ])}})
            mockArtworkFind.mockImplementation(artworkFind)
    
            expect(await getAllCategories("collection")).toMatchSnapshot();
        }
    )

    test.each([
        {
            testName: 'getAllColletions test - throw error when collection not found',
            collectionFind: () => {return {exec: () => Promise.resolve([])}},
            artworkFind: () => {},
            error: "Collection not found"
        },
        {
            testName: 'getAllColletions test - Collection.find() throws error',
            collectionFind: () => {throw Error()},
            artworkFind: () => {},
            error: "Database unavailable"
        },
        {
            testName: 'getAllColletions test - Artwork.find() throws error',
            collectionFind: () => {return {exec: () => Promise.resolve([
                {
                    _id: "6717d46c666e8575d873ee57",
                    name: 'collection',
                    description: 'collection description',
                    __v: 0
                }
                ])}},
            artworkFind: () => {throw Error()},
            error: "Database unavailable"
        },
    ])(`$testName`,
        async ({collectionFind, artworkFind, error}) => {
            mockCollectionFind.mockImplementation(collectionFind)
            mockArtworkFind.mockImplementation(artworkFind)

            expect(() => getAllCategories("collection")).rejects.toThrow(error);
        }
    )
})