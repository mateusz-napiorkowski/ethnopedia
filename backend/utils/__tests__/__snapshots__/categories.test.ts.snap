// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categories util functions tests getAllColletions test - artworks present in collection 1`] = `
[
  "Tytuł",
  "Tytuł.Podtytuł",
  "Tytuł.Podtytuł.Podpodtytuł",
  "Artyści",
  "Rok",
  "Rok.Miesiąc",
  "Rok.Miesiąc.Dzień",
  "Tytuł.Znaczenie tytułu",
  "Rok.Kwartał",
  "Rok.Kwartał.Miesiąc",
  "Rok.Kwartał.Miesiąc.Dzień",
]
`;

exports[`categories util functions tests getAllColletions test - no artworks in collection 1`] = `[]`;
